### PR TITLE
Feature implements the api service for chatgpt

### PR DESCRIPTION
**Pull request by TheoSchlaeppi**
Add ChatGPT API Service to Enable OpenAI API Calls for Book Profile Creation

### Description
This PR introduces a `ChatGPTApiService` class that facilitates basic calls to the OpenAI API, specifically designed to assist with the upcoming tasks related to book profile creation. This service provides a foundational layer to connect with OpenAI’s ChatGPT API, which will be further extended in a separate task to include prompt generation and response parsing.

The current implementation has been manually tested, and API calls are confirmed to work as expected. However, I am encountering challenges in implementing automated tests for this feature and would appreciate feedback or suggestions on how to approach these tests effectively.

- **Added `ChatGPTApiService`**: This service manages requests to OpenAI’s ChatGPT API using Volley.
- **Data Classes**:
  - `ChatGPTMessage`: Represents individual messages with roles (e.g., "user" or "assistant") and content.
  - `ChatGPTRequest`: Structures the model and message list for the API request.
  - `ChatGPTChoice` and `ChatGPTResponse`: Parse response data from the OpenAI API.
- **`sendChatRequest` Function**: Sends user messages as a list, creating a structured API request. Upon receiving the response, it parses the message content, providing success and error handlers.

### Implementation Details
- **Request Queue**: Uses Android’s Volley library to handle network requests.
- **API Key Handling**: Retrieves the API key from `BuildConfig`, depending on the `mode` parameter (production or testing).
- **Error Handling**: Captures and returns both parsing and request errors to provide detailed feedback.
- **Response Parsing**: Extracts the content of the first message choice, ensuring robust parsing even if the response structure varies.

You can review my code already, but I would prefer if you could provide some suggestions for testing this feature first.

### CODE REVIEW

File: app/src/main/java/com/android/bookswap/data/source/api/ChatGPTApiService.kt [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  private val apiUrl = "https://api.openai.com/v1/chat/completions"
+
+  // Send a chat request to the OpenAI API
+  override fun sendChatRequest(
+      userMessages: List<String>,
+      onSuccess: (String) -> Unit,
+      onError: (String) -> Unit
+  ) {
+    val messageList = userMessages.map { m -> ChatGPTMessage(role = "user", content = m) }
+    val chatGPTRequest = ChatGPTRequest(model = "gpt-4o-mini", messages = messageList)
```
Comment (LeniCadorinEpfl) [inline app/src/main/java/com/android/bookswap/data/source/api/ChatGPTApiService.kt:38]: Why did you just the model `"gpt-4o-mini"` and not `"gpt-4"` ?

**Review by LeniCadorinEpfl — CHANGES_REQUESTED** [review]
This PR introduces a well-structured ChatGPTApiService to enable OpenAI API interactions. Here are a few suggestions to further enhance the code and testing:

1. Improving Error Handling with Logging:

Issue: The `sendChatRequest` function uses the onError`` callback to pass errors, but error details are not logged for debugging.
Suggestion: Use Android’s Log.e or Log.w methods to log error messages within onError. This will provide better traceability for request failures, especially during development or in case of unexpected issues.

2.  Mocking Network Requests for Tests:

Issue: Testing the ChatGPTApiService class without network dependency is challenging, making it difficult to automate tests.
Suggestion: Use a dependency injection approach to pass a mock RequestQueue into ChatGPTApiService. This allows using tools like Mockk or Mockito to simulate various API responses (success, error, and malformed JSON) and ensure the onSuccess and onError callbacks function correctly.

3. No Network Error Retrying:

Issue: The current implementation does not retry failed requests in the case of network issues.
Suggestion: Could implement a retry in the case where a request fails.

Overall, great work on implementing this foundational API service! These are just minor suggestions to further enhance robustness and maintainability.

**Comment by TheoSchlaeppi** [comment]
Thank you for the review and valuable suggestions!

- I've added logging within the onError callback as recommended, so error details are now captured in the logs for easier debugging and traceability.
- Regarding network request testing, I explored adding mocks, but found it challenging to implement effectively given the current structure. I may revisit this later to see if there’s a feasible way to mock the requests without excessive complexity.
- As for the retry feature, I agree with your suggestion but believe it would be more manageable to handle retries at a higher level, where we have more control over network conditions and error handling flows.
Thank you again for the feedback, and please let me know if there's anything else!

**Comment by LeniCadorinEpfl** [comment]
Okay, everything looks good to me. Thank you for your work!

**Review by LeniCadorinEpfl — APPROVED** [review]

**Pull request merged**